app/controllers/katello/api/v2/packages_controller.rb

```diff
@@ -81,5 +91,120 @@ def resource_class
     def default_sort
       lambda { |query| query.default_sort }
     end
+
+    def filter_latest(collection)
+      # There are several different ways to implement this:
+
+      # 1) Use a LEFT OUTER JOIN to join `collection` with itself on
+      #    name=name, arch=arch, evr<evr, then select unmatched rows.
+      #    In PostgreSQL 9.2 populated with real data, EXPLAIN shows
+      #    cost=339901.07..381834.02 for the SQL generated by this when `index`
+      #    is called and only a content_view_version_id is specified.
+      #    Unfortunately, this requires writing raw SQL for the JOIN and using
+      #    `collection.to_sql`, which may be somewhat brittle.  For example, be
+      #    aware of quirks like https://github.com/rails/rails/issues/18379
+      collection = collection.joins(
+        "LEFT OUTER JOIN (#{collection.to_sql}) AS katello_rpms2 ON " \
+        'katello_rpms.name = katello_rpms2.name AND katello_rpms.arch = katello_rpms2.arch AND ' \
+        '(CAST(katello_rpms.epoch AS INT) < CAST(katello_rpms2.epoch AS INT) OR ' \
+        ' (CAST(katello_rpms.epoch AS INT) = CAST(katello_rpms2.epoch AS INT) AND ' \
+        '  (katello_rpms.version_sortable COLLATE "C" < katello_rpms2.version_sortable COLLATE "C" OR ' \
+        '   (katello_rpms.version_sortable = katello_rpms2.version_sortable AND ' \
+        '    katello_rpms.release_sortable COLLATE "C" < katello_rpms2.release_sortable COLLATE "C"))))'
+      ).where('katello_rpms2.release_sortable IS NULL')
```

Comment:
apologies again for the delay.  Could we move this to the model?   I think a  'latest' class method would make sense.  It would look somethign like:

def self.latest
 self.joins( "LEFT OUTER JOIN (#{self.to_sql}) ...........
end

Thats really where this belongs. 

Also, what about the large amounts of comments below, planning on getting rid of those?

Reply:
Sure, I can try moving this to the model.

As for the comments:
I selected the uncommented implementation based on the current behavior of Rails and the current performance characteristics of PostgreSQL using the current data set in my Satellite installation.  It is entirely possible that one of the other commented implementations may work significantly better with a different version of Rails and/or with a different version PostgreSQL and/or with a different database and/or with a different data set.
Since the alternative implementations are significantly different from each other, since their advantages/disadvantages are strongly influenced by unspecified implementation details in other pieces of software, and since they are not particularly obvious or intuitive solutions, I thought it would be a good idea to leave them in there as comments so a future developer could easily swap out the implementation if it becomes advantageous to do so.
I can remove them if you want, but we will loose knowledge about those possible alternatives, and someone may have to re-develop them if something changes and causes the uncommented implementation to behave or perform poorly.